In [1]:
import csv


In [2]:
data_collection = []

x_mean = 0
y_mean = 0
x_median = 0
y_median = 0
x_mode = 0
y_mode = 0

x_variance = 0
y_variance = 0
x_st_dev = 0
y_st_dev = 0

xy_covariance = 0
xy_correlation = 0

given_conf_int = []


In [3]:
central_stmt = "{} mean is {} and median is {}."
deviation_stmt = "{} variance is {} and standard deviation is {}."
co_measures_stmt = "{} and {} covariance is {} and correlation is {}."
x_label = ''
y_label = ''


In [16]:
with open('zadanie_korelacja_source.csv') as source_file:
    source_data = csv.reader(source_file, delimiter=',')
        
    for row in source_data:
        try:
            x = float(row[1])
            y = float(row[2])
        except:            
            x_label = row[1]
            y_label = row[2]
        else:
            r = row[0]
            x = float(row[1])
            y = float(row[2])
            z = [r, x, y]
            data_collection.append(z)
            
data_collection


[['x1', 1.3, 76.84878],
 ['x2', 1.95, 81.40732],
 ['x3', 2.466, 77.57895],
 ['x4', 1.6, 74.16341],
 ['x5', 2.952, 68.84907],
 ['x6', 1.801, 73.88595],
 ['x7', 1.34, 75.26829],
 ['x8', 1.34, 76.26098],
 ['x9', 1.63, 80.57244],
 ['x10', 2.326, 65.46259],
 ['x11', 2.436, 67.5482],
 ['x12', 6.563, 55.02451],
 ['x13', 1.283, 76.2799],
 ['x14', 1.43, 82.29024],
 ['x15', 1.988, 69.80695],
 ['x16', 1.61, 81.40112],
 ['x17', 1.921, 82.19756],
 ['x18', 2.4, 74.22683],
 ['x19', 1.495, 80.12888],
 ['x20', 4.705, 45.55095],
 ['x1', 1.3, 76.84878],
 ['x2', 1.95, 81.40732],
 ['x3', 2.466, 77.57895],
 ['x4', 1.6, 74.16341],
 ['x5', 2.952, 68.84907],
 ['x6', 1.801, 73.88595],
 ['x7', 1.34, 75.26829],
 ['x8', 1.34, 76.26098],
 ['x9', 1.63, 80.57244],
 ['x10', 2.326, 65.46259],
 ['x11', 2.436, 67.5482],
 ['x12', 6.563, 55.02451],
 ['x13', 1.283, 76.2799],
 ['x14', 1.43, 82.29024],
 ['x15', 1.988, 69.80695],
 ['x16', 1.61, 81.40112],
 ['x17', 1.921, 82.19756],
 ['x18', 2.4, 74.22683],
 ['x19', 1.495, 80.1

In [15]:
def mean(input_list, column):
    result = 0
    collection = []
    
    for row in input_list:
        collection.append(row[column])
            
    result = sum(collection) / len(collection)
    
    return result


mean(data_collection, 1)


2.2268

In [17]:
def median(input_list, column):
    result = 0
    collection = []
    input_parity = 0
    sample_size = 0
    
    for row in input_list:
        collection.append(row[column])
    
    collection.sort()    
    input_parity = len(collection) % 2
    sample_size = len(collection)
    
    if input_parity == 0:
        a = int(sample_size / 2 - 1)
        b = int(sample_size / 2)
        result = (collection[a] + collection[b]) / 2
    else:
        c = int(sample_size / 2 + 0.5)
        result = collection[c]
        
    return result


median(data_collection, 1)


1.861

In [7]:
def mode(input_list, column):
    print('tu będxzie funkcjsa mody')


In [18]:
def variance(input_list, column):
    collection = []
    deviances = []
    result = 0
    sample_mean = 0
    
    for row in input_list:
        collection.append(row[column])
        
    sample_mean = sum(collection) / len(collection)
    
    for i in collection:
        a = (i - sample_mean) ** 2
        deviances.append(a)
            
    result = sum(deviances) / (len(collection) - 1)
    
    return result


variance(data_collection, 1)


1.6226544205128206

In [19]:
def st_dev(input_list, column):
    result = 0
    
    result = variance(input_list, column) ** (1/2)
    
    return result


st_dev(data_collection, 1)


1.2738345341969735

In [20]:
def covariance(input_list):
    x_mean = mean(input_list, 1)
    y_mean = mean(input_list, 2)
    deviances = []
    result = 0
    
    for i in input_list:
        a = (i[1] - x_mean) * (i[2] - y_mean)
        deviances.append(a)
        
    result = sum(deviances) / (len(deviances) - 1)
       
    return result


covariance(data_collection)


-9.688737665948716

In [21]:
def correlation(input_list):
    result = 0
    
    result = covariance(input_list) / (st_dev(input_list, 1) * st_dev(input_list, 2))
    
    return result


correlation(data_collection)


-0.8204930122649573

In [22]:
def conf_interval(input_list, column, pop_st_dev, level):
    collection = []
    z = 0
    factor = 0
    result = ()
    sample_mean = 0
    
    for row in input_list:
        collection.append(row[column])
        
    sample_mean = sum(collection) / len(collection)
    
    if level == 99:
        z = 2.576
    elif level == 98:
        z = 2.326
    elif level == 95:
        z = 1.96
    elif level == 90:
        z = 1.645
    else:
        print("Wrong confidence level provided")
        
    factor = z * (pop_st_dev / (len(data_collection) ** (1/2)))
        
    result = ((sample_mean - factor), (sample_mean + factor))
    
    return result


conf_interval(data_collection, 2, 9.09, 95)


(70.42062581476881, 76.05466618523121)

In [13]:
y_mean = mean(data_collection, 2)
y_median = median(data_collection, 2)

y_variance = variance(data_collection, 2)
y_st_dev = st_dev(data_collection, 2)

xy_covariance = covariance(data_collection)
xy_correlation = correlation(data_collection)

given_conf_int = conf_interval(data_collection, 2, 9.09, 95)


In [14]:
print(central_stmt.format(y_label, "{:.2f}".format(y_mean),
                          "{:.2f}".format(y_median)))                      
print(deviation_stmt.format(y_label, "{:.2f}".format(y_variance),
                            "{:.2f}".format(y_st_dev)))
print(co_measures_stmt.format(x_label, y_label, "{:.2f}".format(xy_covariance),
                              "{:.2f}".format(xy_correlation)))
print("{} 95% confidence interval for its mean is from {} to {}.".format(y_label,
                                                                     "{:.2f}".format(given_conf_int[0]),
                                                                     "{:.2f}".format(given_conf_int[1])))

Life expectancy at birth mean is 73.24 and median is 75.76.
Life expectancy at birth variance is 88.19 and standard deviation is 9.39.
Fertility rate and Life expectancy at birth covariance is -9.94 and correlation is -0.82.
Life expectancy at birth 95% confidence interval for its mean is from 69.25 to 77.22.


# Statistics for data provided

Life expectancy at birth mean is 73.24 and median is 75.76.
Life expectancy at birth 95% confidence interval for its mean is from 69.25 to 77.22.
Life expectancy at birth variance is 88.19 and standard deviation is 9.39.
Fertility rate and Life expectancy at birth covariance is -9.94 and correlation is -0.82.

Considering above figures, very strong negative correlation between Fertility rate and Life expectancy at birth has been observed. Such a correlation could be explained by a level of well-being in the countries observed, but further investigation to confirm this hypothesis is needed.
